src/code/z_view.c

```diff
@@ -279,137 +279,231 @@ void View_RenderView(View* view, s32 uParm2) {
     }
 }
 
-#ifdef NON_MATCHING
-// saved register usage is wrong, stack, regalloc
 s32 View_RenderToPerspectiveMatrix(View* view) {
     f32 aspect;
     s32 width;
     s32 height;
     Vp* vp;
     Mtx* projection;
     Mtx* viewing;
-    GraphicsContext* gfxCtx;
+    GraphicsContext* gfxCtx = view->gfxCtx;
 
-    gfxCtx = view->gfxCtx;
+    OPEN_DISPS(gfxCtx);
 
-    {
-        Vp* _vp = (Vp*)gfxCtx->polyOpa.d - 1;
-        vp = _vp;
-        gfxCtx->polyOpa.d = (Gfx*)_vp;
-    }
+    vp = GRAPH_ALLOC(gfxCtx, sizeof(*vp));
     View_ViewportToVp(vp, &view->viewport);
     view->vp = *vp;
 
     View_SetScissorForLetterbox(view);
 
-    gSPViewport(gfxCtx->polyOpa.p++, vp);
-    gSPViewport(gfxCtx->polyXlu.p++, vp);
+    gSPViewport(POLY_OPA_DISP++, vp);
+    gSPViewport(POLY_XLU_DISP++, vp);
 
-    {
-        Mtx* _m = (Mtx*)gfxCtx->polyOpa.d - 1;
-        projection = _m;
-        gfxCtx->polyOpa.d = (Gfx*)_m;
-    }
+    projection = GRAPH_ALLOC(gfxCtx, sizeof(*projection));
     view->projectionPtr = projection;
 
     width = view->viewport.rightX - view->viewport.leftX;
     height = view->viewport.bottomY - view->viewport.topY;
     aspect = (f32)width / (f32)height;
 
     guPerspective(projection, &view->normal, view->fovy, aspect, view->zNear, view->zFar, view->scale);
-
     view->projection = *projection;
+    //! @bug: This cast of `projection` is invalid
+    View_StepQuake(view, (RSPMatrix*)projection);
```

Comment:
This is very weird. I think `View_StepQuake` definitely takes an `RSPMatrix *`. It calls `SysMatrix_FromRSPMatrix` & `SysMatrix_GetStateAsRSPMatrix` which aren't decompiled yet but make much more sense with `RSPMatrix` than `Mtx`.

If `view->projection` was an `RSPMatrix` instead of an `Mtx`, that would mean `guPerspective`, `guOrtho`, etc. would have to operate on `RSPMatrix` -- which doesn't seem to work out either?

Reply:
`RSPMatrix` and `Mtx` are actually the exact same type, but iirc some code required certain accesses that `Mtx` could not accommodate. The matrix types are a bit of a mess right now, these two types should probably be a union ideally, since they represent the exact same underlying data. This is likely best addressed in a separate PR however, since all usages would need cleaning up.